fdbserver/Resolver.actor.cpp

```diff
@@ -50,6 +52,61 @@ struct ProxyRequestsInfo {
 	ProxyRequestsInfo() = default;
 };
 
+class RecentStateTransactionsInfo {
+public:
+	RecentStateTransactionsInfo() = default;
+
+	// Erases state transactions up to the given version (inclusive) and returns
+	// the number of bytes for the erased mutations.
+	int64_t eraseUpTo(Version oldestVersion) {
+		recentStateTransactions.erase(recentStateTransactions.begin(),
+		                              recentStateTransactions.upper_bound(oldestVersion));
+
+		int64_t stateBytes = 0;
+		while (recentStateTransactionSizes.size() && recentStateTransactionSizes.front().first <= oldestVersion) {
+			stateBytes += recentStateTransactionSizes.front().second;
+			recentStateTransactionSizes.pop_front();
+		}
+		return stateBytes;
+	}
+
+	// Adds state transactions between two versions to the reply message.
+	// Returns if shardChanged has ever happened for these versions.
+	bool addStateTransactions(ResolveTransactionBatchReply* reply, Version firstUnseenVersion, Version commitVersion) {
+		bool shardChanged = false;
+		auto stateTransactionItr = recentStateTransactions.lower_bound(firstUnseenVersion);
+		auto endItr = recentStateTransactions.lower_bound(commitVersion);
```

Comment:
It is possible, `endItr` points to the current commit version, which has state mutations thus not the `end()`.

Reply:
If there is a state txn in the same batch, will this cause broadcast as well? My understanding is yes.

If so, should we use upper_bound here? We can also remove line 377 when using upper_bound, IIUC. So that there is only one place to calculate shardChanged.

Reply:
Not necessarily. There are the state txns that are not related to shard changes.

I refactored the code so that computing `shardChanged` happened within this function.

Reply:
What I wanted to ask is, `If there is a shard changing state txn in the same batch, will this cause broadcast as well?`

What I really want to ask is, why processing the shard change for the current batch differently than earlier shard change txn? It seems that using `upper_bound` in line 81 and 82 should work, so that all the shard changing txns at and before `commitVersion` are processed in the same way. Unless my understanding is wrong.

Reply:
> What I wanted to ask is, `If there is a shard changing state txn in the same batch, will this cause broadcast as well?`

Yes. But having state txn doesn't necessarily mean there is a shard change, which can only be determined from `toCommit.isShardChanged();`.

> 
> What I really want to ask is, why processing the shard change for the current batch differently than earlier shard change txn? It seems that using `upper_bound` in line 81 and 82 should work, so that all the shard changing txns at and before `commitVersion` are processed in the same way. Unless my understanding is wrong.

Using `upper_bound` won't work, because for the `commitVersion`, we do *NOT* add state txns back to the reply. This is different from how `shardChanged`, which should include `commitVersion`.

Reply:
Thank, got it. As chatted, the key difference is that state txn in the same batch is processed differently than state txn in earlier batch. It'll be great if we can document this.